Make HTTP frames a TLV format. This has several advantages:

ekr opened the pull request:
I know this is a late change, but I think it's an improvement....

- You can switch on the frame type before you even read a
  variable-length value
- You can have special-case (negotiated) frame types with no lengths

If this has already been discussed and rejected somewhere, maybe
I missed it.

ekr commented:
@DavidSchinazi

mikkelfj commented:
Me thinks it makes sense. As to 0-length special DATA frames (until end), I also think it could be a length-less separate type instead of 0-length. The downside is that you can't iterate frames without understanding all types, but that is no different from transport.

ekr commented:
I think the way you would make iterating all frames possible is by requiring explicit opt-in (or standards or something) for any types that didn't have the TLV format.

DavidSchinazi commented:
I like this. You could imagine reserving a part of the type space (e.g. 0-32) for non-TLV frames, and if you ever receive something in that range that you don't understand you fail the connection. But if you receive a TLV type you don't understand, you can safely ignore it since you know how long it is.

Minor: I'd make the type a variable-length integer

ekr commented:
I would be fine with type being a VLI.

On Thu, Dec 20, 2018 at 3:57 PM David Schinazi <notifications@github.com>
wrote:

> I like this. You could imagine reserving a part of the type space (e.g.
> 0-32) for non-TLV frames, and if you ever receive something in that range
> that you don't understand you fail the connection. But if you receive a TLV
> type you don't understand, you can safely ignore it since you know how long
> it is.
>
> Minor: I'd make the type a variable-length integer
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/quicwg/base-drafts/pull/2235#issuecomment-449181514>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABD1oVtlQi7p7siJy9Y9WKwIWjwPPuL6ks5u7CPmgaJpZM4ZdEz2>
> .
>

dtikhonov commented:
We've had this discussion haven't we.

mikkelfj commented:
L=0 definitely was discussed in the context of whether only DATA frames or all frames could be 0 length. But that is a different issue.

martinthomson commented:
This has been discussed a few times.  We discussed this in h2 and explicitly made the opposite choice.  Though that was a while back, and I don't know if it was well documented.

ianswett commented:
This seems like a good change to me, I was rather confused when I realized the length came before the type.  If we're really going to drink the varint kool-aid, maybe the type should be a varint too?

kazuho commented:
I'm fine with moving frame type in front of the length.

OTOH, I am not sure if I prefer using varints for frame types.

IIRC, using varints for frame types of the transport, with the requirement that the minimal-sized encoding MUST be used, was based on the fact that it reserves more code points to future extensions, at the same time not making existing implementations complex. The reason the change did not make existing implementations complex was because new frame types cannot be used without negotiation.

The situation is different in H3.

H3 allows frames to be used without negotiation. In fact, we allow endpoints to grease the codepoints.

That means that every implementation would be required to support frame types using varints, even though in practice only the single-octet versions would be used.

To me that sounds like an unnecessary complexity.

I am also skeptical if we would ever run out of H3 frame types even if they are kept 8 bits, considering how few we have used in H2.

dtikhonov commented:
I don't see an advantage to having type come before length.  I would be against special-casing some frames to have no length, which is the only scenario where this change would be necessary for.

HTTP/2 frames have length before type; why change?  Seems gratuitous.

mikkelfj commented:
Nearly all binary self-describing formats introduced these days use some form of TLV because it offers the a good size/flexibility tradeoff, because the L and/or V part can omitted or specialised. For example there can be compact integers up to 4 bits in length using only a type field and a separate var length type. That is not in itself an argument for using it in HTTP/3.

LPardue commented:
I'm not sure about omitting length. I can see there could be benefit in stating that some frame types, that are part of H3 core, which contain only varints could be optimised by not including the length. 

However, from an implementation perspective it seems simpler to have code that can read frames out of the stream using the frame length.

dtikhonov commented:
The implementation is easy either way, which is why I ask: "why change?" The current frame format derives from HTTP/2, which makes me inclined to leave it as-is.

ekr commented:
On Mon, Dec 24, 2018 at 7:07 AM Dmitri Tikhonov <notifications@github.com>
wrote:

> The implementation is easy either way, which is why I ask: "why change?"
>
I listed the arguments for this in the original comments on the issue.

dtikhonov commented:
@ekr, I addressed the original comments above. The proposed change is only advantageous if we want to be able to special-case some frame types to exclude length. We should get a consensus on that first.

ekr commented:
No, I don't think we need consensus on that first, because (for instance) someone might want a private extension that permitted excluding the length, which the current design makes very difficult. TLV is just inherently more flexible. And so far the only argument I've heard against it is "why change".

dtikhonov commented:
Excluding the length has diminished benefit if it takes 8 bytes to encode the frame type (I am referring to #2253, with which this proposal is linked).

I am not against future-proofing the format, but both this and the varint type proposal seem very theoretical: _there might be someone who may want..._ etc.  For example, there have only been [two extensions](https://www.iana.org/assignments/http2-parameters/http2-parameters.xhtml#frame-type) to the HTTP/2 frames.  Has extending or experimenting with HTTP/2 frames ever been a real problem?

ianswett commented:
After thinking about this more, I think this is a very important change and making it will cause HTTP/3 align better with the QUIC transport.

In retrospect, PR#2098(Allow DATA to extend to the end of the stream) violated what appears to be a core design choice of HTTP/2, that all frames explicitly specify a length, so they're easily skippable.  Admittedly, skipping DATA doesn't make much sense, but it is still inconsistent.

To me, that's one of many signals that TLV is a better approach.  I was intending to create an extension that was DATA without length, but I realized that this new frame type would have a length byte, but it would be ignored or specified as a particular incorrect value(ie: 0), neither of which are ideal.

dtikhonov commented:
>  this is a very important change and making it will cause HTTP/3 align better with the QUIC transport.

1. What is meant by "better alignment?"
1. Why is it important?  QUIC and HTTP/3 are different protocols, we touched on this a bit in #2232.

ianswett commented:
@dtikhonov By better alignment, I meant more consistent design choices between QUIC transport and HTTP/3.

Previously HTTP over QUIC adopted varints because it was straightforward and QUIC transport had recently adopted them.  Moving HTTP/3 to TLV seems like a similar change.

More importantly, I'd like to argue that TLV is fundamentally better than LTV.  If someone can provide an example in HTTP/2 where LTV has shown clear value over TLV, I'd be interested in hearing about it.

dtikhonov commented:
> Previously HTTP over QUIC adopted varints because it was straightforward and QUIC transport had recently adopted them. Moving HTTP/3 to TLV seems like a similar change.

One does not follow from the other; TLV vs LTV and fixed-size vs varint are two different, orthogonal questions.

> If someone can provide an example in HTTP/2 where LTV has shown clear value over TLV, I'd be interested in hearing about it.

I'd be curious to know why HTTP/2 chose one over the other in the first place.

ekr commented:
On Mon, Dec 24, 2018 at 11:11 AM Dmitri Tikhonov <notifications@github.com>
wrote:

> Previously HTTP over QUIC adopted varints because it was straightforward
> and QUIC transport had recently adopted them. Moving HTTP/3 to TLV seems
> like a similar change.
>
> One does not follow from the other; TLV vs LTV and fixed-size vs varint
> are two different, orthogonal questions.
>

This is true at some level, but once we've changed anything, we might as
well ask what the best design is.

ianswett commented:
The varint question is completely orthogonal, sorry to bring it up here. @martinthomson filed a separate issue for that: #2253

DavidSchinazi commented:
We could have h3 use both TV and TLV: for example we say that
- all even types are TV, and must be negotiated before use (or in base spec)
- all odd types are TLV, and can be used without negotiation and ignored on receipt if unknown

This allows efficient encoding while preserving extensibility.

mikkelfj commented:
let's not over-engineer more than we already have

janaiyengar commented:
I wasn't in the room where it happened, but I can imagine one reason for doing LTV in HTTP/2 might have been that you need to read out frames from the underlying byte-stream before demuxing and parsing them. A frame reader only needs Length before reading out the entire frame and passing it on to the parser. QUIC is different in that the entire frame is present when the length is read.

I think this is a simple enough change, and there are straightforward benefits. We expect to see some changes with more implementation experience in the HTTP mapping, so I'd like to push for this PR.

@ekr -- this discussion should be happening on an issue. Can you open an issue ?

dtikhonov commented:
@janaiyengar writes, in part (emphasis added):
> QUIC is different in that the _entire frame is present_ when the length is read.

I don't understand what you mean by that.

LPardue commented:
Yeah I don't follow either. The h3 layer is presented an in-order byte stream for each stream but it still needs to parse out H3 frames. While there are some loose assurances on frame ordering I don't know it QUIC is so different from H2, even with separate streams

janaiyengar commented:
@dtikhonov, @LPardue: Urgh, yes, sorry, I was not thinking straight. I was incorrect.

LTV or TLV, I don't have a strong preference either way.

martinthomson commented:
I interpreted @janaiyengar's comment to mean that when you read a QUIC frame it's in a datagram, so it's probably all sitting in memory.  With a streamed frame, you might not have all of the frame in place.  You might only have the length.

LPardue commented:
Yeah I think we are all in agreement on that.

MikeBishop commented:
The previous discussion was #1351.  The conclusion there was that the order doesn't matter a great deal (under the assumption that you need both), and so consistency with HTTP/2 was the tie-break.

If there's a desire to enable length-less frames, that's an argument for TLV.  However, I don't think switching on the type value first is terribly compelling -- there doesn't seem to be a great difference between:
- passing your frame parser the type and expecting it to tell you how many bytes of the stream it processed and whether the frame was valid
- passing your frame parser the type and length and expecting it to tell you if the frame was mis-sized

janaiyengar commented:
@martinthomson 
> I interpreted @janaiyengar's comment to mean that when you read a QUIC frame it's in a datagram, so it's probably all sitting in memory. With a streamed frame, you might not have all of the frame in place. You might only have the length.

Yes, that's exactly what I meant, but this discussion is about HTTP frames, which is what I had confused in my mind.

mikkelfj commented on draft-ietf-quic-http.md:
```diff
@@ -402,22 +402,23 @@ All frames have the following format:
   Length:
   : A variable-length integer that describes the length of the Frame Payload.
     This length does not include the Type field.
 
```
With reordering the "Length" description no longer makes sense. If anything, length should state that it does not include itself.
